streams/src/main/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreator.java

```diff
@@ -0,0 +1,237 @@
+                      final StreamsMetricsImpl streamsMetrics,
+                      final StateDirectory stateDirectory,
+                      final ChangelogReader storeChangelogReader,
+                      final ThreadCache cache,
+                      final Time time,
+                      final KafkaClientSupplier clientSupplier,
+                      final String threadId,
+                      final Logger log) {
+        applicationId = config.getString(StreamsConfig.APPLICATION_ID_CONFIG);
+        this.builder = builder;
+        this.config = config;
+        this.streamsMetrics = streamsMetrics;
+        this.stateDirectory = stateDirectory;
+        this.storeChangelogReader = storeChangelogReader;
+        this.time = time;
+        this.log = log;
+
+        if (EXACTLY_ONCE.equals(config.getString(StreamsConfig.PROCESSING_GUARANTEE_CONFIG))) {
+            threadProducer = null;
+            taskProducers = new HashMap<>();
+        } else {
+            final String threadProducerClientId = getThreadProducerClientId(threadId);
+            final Map<String, Object> producerConfigs = config.getProducerConfigs(threadProducerClientId);
+            log.info("Creating thread producer client");
+            threadProducer = clientSupplier.getProducer(producerConfigs);
+            taskProducers = Collections.emptyMap();
+        }
+
+
+        this.cache = cache;
+        this.threadId = threadId;
+        this.clientSupplier = clientSupplier;
+
+        createTaskSensor = ThreadMetrics.createTaskSensor(threadId, streamsMetrics);
+    }
+
+    Collection<Task> createTasks(final Consumer<byte[], byte[]> consumer,
+                                 final Map<TaskId, Set<TopicPartition>> tasksToBeCreated) {
+        final List<Task> createdTasks = new ArrayList<>();
+        for (final Map.Entry<TaskId, Set<TopicPartition>> newTaskAndPartitions : tasksToBeCreated.entrySet()) {
+            final TaskId taskId = newTaskAndPartitions.getKey();
+            final Set<TopicPartition> partitions = newTaskAndPartitions.getValue();
+
+            final String threadIdPrefix = String.format("stream-thread [%s] ", Thread.currentThread().getName());
+            final String logPrefix = threadIdPrefix + String.format("%s [%s] ", "task", taskId);
+            final LogContext logContext = new LogContext(logPrefix);
+
+            final ProcessorTopology topology = builder.buildSubtopology(taskId.topicGroupId);
+
+            final ProcessorStateManager stateManager = new ProcessorStateManager(
+                taskId,
+                partitions,
+                Task.TaskType.ACTIVE,
+                stateDirectory,
+                topology.storeToChangelogTopic(),
+                storeChangelogReader,
+                logContext
+            );
+
+            if (threadProducer == null) {
+                // create one producer per task for EOS
+                // TODO: after KIP-447 this would be removed
```

Comment:
I don't think we need to keep this TODO, as only after a stream 3.0 is there, we shall remove the support for task producer.